python/cudf/cudf/utils/cudautils.py

```diff
@@ -113,20 +114,34 @@ def gpu_copy(inp, out):
 
 
 def astype(ary, dtype):
-    if ary.dtype == np.dtype(dtype):
+
+    dtype = np.dtype(dtype)
+
+    if ary.dtype == dtype:
         return ary
-    elif (
-        ary.dtype == np.int64 or ary.dtype == np.dtype("datetime64[ms]")
-    ) and (dtype == np.dtype("int64") or dtype == np.dtype("datetime64[ms]")):
-        return ary.view(dtype)
-    elif ary.size == 0:
+
+    # if converting to or from a datetime *and* to or from
+    # an integer array of the same width, return a view of
+    # the input array of the output dtype
+    to_datetime = dtype.type is np.datetime64
+    if to_datetime or ary.dtype.type is np.datetime64:
+        # this will the datetime dtype
+        a = dtype if to_datetime else ary.dtype
+        # this will be the opposite dtype
+        b = ary.dtype if to_datetime else dtype
+        # if the datetime's numeric dtype matches the
+        # other dtype, it's safe to cast between them
+        if datetime_to_numerical_dtype(a) == b:
+            return ary.view(dtype)
```

Comment:
This isn't the expected behavior of typecasting between datetime types IIRC, it expects the equivalent datetime value in the requested resolution, so you'd just need to multiply or divide by the conversion factor.

Reply:
Makes sense. I added [logic to scale the values](https://github.com/rapidsai/cudf/pull/2213/commits/9febb0ebbc2cef5228b5af581541917efa169f4d#diff-cfbfb41f0663e4931eb49ed66e4c3a34R175) when casting between datetimes with different resolutions in the latest push. I also moved this logic to the `DatetimeColumn` b/c I don't think we should ever encounter a source datetime DeviceArray and a dst datetime dtype in this method.